api/src/main/java/org/pmiops/workbench/actionaudit/ActionAuditQueryServiceImpl.java

```diff
@@ -0,0 +1,98 @@
+package org.pmiops.workbench.actionaudit;
+
+import com.google.cloud.bigquery.FieldValueList;
+import com.google.cloud.bigquery.QueryJobConfiguration;
+import com.google.cloud.bigquery.TableResult;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import java.util.List;
+import java.util.stream.StreamSupport;
+import javax.inject.Provider;
+import org.pmiops.workbench.api.BigQueryService;
+import org.pmiops.workbench.config.WorkbenchConfig;
+import org.pmiops.workbench.config.WorkbenchConfig.ActionAuditConfig;
+import org.pmiops.workbench.model.AuditLogEntriesResponse;
+import org.pmiops.workbench.model.AuditLogEntry;
+import org.pmiops.workbench.utils.FieldValues;
+import org.springframework.stereotype.Service;
+
+@Service
+public class ActionAuditQueryServiceImpl implements ActionAuditQueryService {
+
+  private BigQueryService bigQueryService;
+  private Provider<WorkbenchConfig> workbenchConfigProvider;
+  private static final String WORKSPACE_EVENTS_QUERY_STRING_FORMAT =
+      "SELECT\n"
+          + "  TIMESTAMP_MILLIS(CAST(jsonPayload.timestamp AS INT64)) as event_time,\n"
+          + "  jsonPayload.agent_type AS agent_type,\n"
+          + "  CAST(jsonPayload.agent_id AS INT64) AS agent_id,\n"
+          + "  jsonPayload.agent_email AS agent_username,\n"
+          + "  jsonPayload.action_id AS action_id,\n"
+          + "  jsonPayload.action_type AS action_type,\n"
+          + "  jsonPayload.target_type AS target_type,\n"
+          + "  CAST(jsonPayload.target_id AS INT64) AS target_id,\n"
+          + "  jsonPayload.target_property AS target_property,\n"
+          + "  jsonPayload.prev_value AS prev_value,\n"
+          + "  jsonPayload.new_value AS new_value\n"
+          + "FROM %s\n"
+          + "WHERE jsonPayload.target_id = %d AND\n"
+          + "  jsonPayload.target_type = 'WORKSPACE'\n"
+          + "ORDER BY event_time, agent_id, action_id\n"
+          + "LIMIT %d";
+
+  public ActionAuditQueryServiceImpl(
+      BigQueryService bigQueryService, Provider<WorkbenchConfig> workbenchConfigProvider) {
+    this.bigQueryService = bigQueryService;
+    this.workbenchConfigProvider = workbenchConfigProvider;
+  }
+
+  @Override
+  public AuditLogEntriesResponse queryEventsForWorkspace(long workspaceDatabaseId, long limit) {
```

Comment:
You mean at the service level? Sure. Maybe 1000 would work. Beyond that we'd need pagination, which is simple here since it only grows by appending.

Auditable events are relatively rare, and the ending should be restricted to admin users.